src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,308 @@
+  smul_with_zero α (hamm β) := pi.smul_with_zero _
+instance [Π i, add_monoid (β i)] : add_monoid (hamm β) := pi.add_monoid
+instance [Π i, add_comm_monoid (β i)] : add_comm_monoid (hamm β) := pi.add_comm_monoid
+instance [Π i, add_comm_group (β i)] : add_comm_group (hamm β) := pi.add_comm_group
+instance (α) [semiring α] (β : ι → Type*) [Π i, add_comm_monoid (β i)]
+  [Π i, module α (β i)] : module α (hamm β) := pi.module _ _ _
+
+/- API to/from the type synonym. -/
+
+/-- `to_hamm` is the identity function to the `hamm` of a type.  -/
+@[pattern] def to_hamm : (Π i, β i) ≃ hamm β := equiv.refl _
+
+/-- `of_hamm` is the identity function from the `hamm` of a type.  -/
+@[pattern] def of_hamm : hamm β ≃ Π i, β i := equiv.refl _
+
+@[simp] lemma to_hamm_symm_eq                 : (@to_hamm _ β).symm = of_hamm := rfl
+@[simp] lemma of_hamm_symm_eq                 : (@of_hamm _ β).symm = to_hamm := rfl
+@[simp] lemma to_hamm_of_hamm (x : hamm β)    : to_hamm (of_hamm x) = x := rfl
+@[simp] lemma of_hamm_to_hamm (x : Π i, β i)  : of_hamm (to_hamm x) = x := rfl
+@[simp] lemma to_hamm_inj {x y : Π i, β i}    : to_hamm x = to_hamm y ↔ x = y := iff.rfl
+@[simp] lemma of_hamm_inj {x y : hamm β}      : of_hamm x = of_hamm y ↔ x = y := iff.rfl
+
+@[simp] lemma to_hamm_zero [Π i, has_zero (β i)] : to_hamm (0 : Π i, β i) = 0 := rfl
+@[simp] lemma of_hamm_zero [Π i, has_zero (β i)] : of_hamm (0 : hamm β) = 0 := rfl
+@[simp] lemma to_hamm_one [Π i, has_one (β i)] : to_hamm (1 : Π i, β i) = 1 := rfl
+@[simp] lemma of_hamm_one [Π i, has_one (β i)] : of_hamm (1 : hamm β) = 1 := rfl
+@[simp] lemma to_hamm_neg [Π i, has_neg (β i)] {x : Π i, β i} : to_hamm (-x) = - to_hamm x := rfl
+@[simp] lemma of_hamm_neg [Π i, has_neg (β i)] {x : hamm β} : of_hamm (-x)  = - of_hamm x := rfl
+@[simp] lemma to_hamm_add [Π i, has_add (β i)] {x y : Π i, β i} :
+  to_hamm (x + y) = to_hamm x + to_hamm y := rfl
+@[simp] lemma of_hamm_add [Π i, has_add (β i)] {x y : hamm β} :
+  of_hamm (x + y) = of_hamm x + of_hamm y := rfl
+@[simp] lemma to_hamm_sub [Π i, has_sub (β i)] {x y : Π i, β i} :
+  to_hamm (x - y) = to_hamm x - to_hamm y := rfl
+@[simp] lemma of_hamm_sub [Π i, has_sub (β i)] {x y : hamm β} :
+  of_hamm (x - y) = of_hamm x - of_hamm y := rfl
+@[simp] lemma to_hamm_mul [Π i, has_mul (β i)] {x y : Π i, β i} :
+  to_hamm (x * y) = to_hamm x * to_hamm y := rfl
+@[simp] lemma of_hamm_mul [Π i, has_mul (β i)] {x y : hamm β} :
+  of_hamm (x * y) = of_hamm x * of_hamm y := rfl
+@[simp] lemma to_hamm_smul [Π i, has_smul α (β i)] {r : α} {x : Π i, β i} :
+  to_hamm (r • x) = r • to_hamm x := rfl
+@[simp] lemma of_hamm_smul [Π i, has_smul α (β i)] {r : α} {x : hamm β} :
+  of_hamm (r • x) = r • of_hamm x := rfl
+
+section
+
+/- We define the instances relating to norm and dist. -/
+
+variables [fintype ι] [Π i, decidable_eq (β i)]
+
+instance : has_dist (hamm β)    := ⟨λ x y, hamm_dist (of_hamm x) (of_hamm y)⟩
+instance : has_nndist (hamm β) := ⟨λ x y, hamm_dist (of_hamm x) (of_hamm y)⟩
```

Comment:
I don't think you want this instance, it will form a diamond with [`pseudo_metric_space.to_has_nndist`](https://leanprover-community.github.io/mathlib_docs/topology/metric_space/basic.html#pseudo_metric_space.to_has_nndist)

Reply:
But `has_nnnorm` is fine? Or will that also form a diamond?

Reply:
Oh, that should go too

Reply:
```
example : @has_nndist.nndist (hamm β) (@pseudo_metric_space.to_has_nndist (hamm β) _) = @has_nndist.nndist (hamm β) (hamm.has_nndist) := rfl
```

This seems fine?

Reply:
`has_nnnorm` will need a weakening of the lemma but I think it'll be OK.